### PR TITLE
Allow specifying a device checksum in the metainfo.xml file

### PR DESCRIPTION
**Pull request by hughsie**
Some firmware has a different on-device checksum to the hash of the firmware
file itself. This may be because:

 * The content is not a binary file, e.g. Intel HEX or SREC
 * Only part of the firmware is flashed, e.g. ignoring the bootloader section
 * The device checksum is calculated using another method entirely, e.g. PCR0

### CODE REVIEW

File: app/metadata.py [hunk 1]
```diff
@@ -160,6 +160,13 @@ def _generate_metadata_kind(filename, fws, firmware_baseuri=''):
                 csum.set('filename', md.filename_contents)
                 csum.set('target', 'content')
 
+            # add device checksum
+            if md.checksum_device:
+                csum = ET.SubElement(rel, 'checksum')
+                csum.text = md.checksum_device
+                csum.set('type', 'sha1')
```
Comment (superm1) [inline app/metadata.py]: sha1? Since this is new I feel it should be sha256

File: app/views_upload.py [hunk 2]
```diff
@@ -133,6 +133,12 @@ def _create_fw_from_uploaded_file(ufile):
         md.checksum_contents = csum.get_value()
         md.filename_contents = csum.get_filename()
 
+        # from the device checksum
+        if hasattr(AppStreamGlib.ChecksumTarget, 'DEVICE'):
```
Comment (superm1) [inline app/views_upload.py]: Weren't you trying to remove the need for appstream glib in lvfs-website?

File: app/metadata.py (hunk 1 above)
Comment (hughsie) [reply app/metadata.py]: We should really move to a better hash format at some point. We do need to support /both/ as some hardware will only spit out SHA1 for the PCR0 case. I think supporting "both" means "any", so it's quite a bigger change to rearchitect the databases to split out the checksums to a different table and to add a proper `Checksum` object in `models.py`. Certainly needs doing, but not today.

File: app/views_upload.py (hunk 2 above)
Comment (hughsie) [reply app/views_upload.py]: I think here it makes sense still in `uploadedfile.py`; we can re-use a lot of the validation stuff from the library without re-implementing all the rules in python.

File: app/metadata.py (hunk 1 above)
Comment (superm1) [reply app/metadata.py]: As long as from the OEM side building the metainfo we can just fill in both now without LVFS getting upset pushing that back is fine.

**Comment by hughsie** [comment]
Looks good now? It's getting complicated juggling PRs :)

**Pull request merged**